lib/directory-view.coffee

```diff
@@ -74,6 +68,20 @@ class DirectoryView
     @element.entries = @entries
     @element.directoryName = @directoryName
 
+  updateIcon: ->
+    if service = IconServices.get 'element-icons'
+      @subscriptions.add service @directoryName, @directory.path
+    else
+      if @directory.symlink
+        iconClass = 'icon-file-symlink-directory'
+      else
+        iconClass = 'icon-file-directory'
+        if @directory.isRoot
+          iconClass = 'icon-repo' if repoForPath(@directory.path)?.isProjectAtRoot()
+        else
+          iconClass = 'icon-file-submodule' if @directory.submodule
+      @directoryName.classList.add(iconClass)
+
```

Comment:
@nathansobo Concerning [your recommendation](https://github.com/atom/tree-view/pull/1146#issuecomment-339709171) with the `updateIcon` method... how do you propose I handle this logic? It sounded like you wanted the bulk of the class-assignment logic handled in the `TreeViewPackage` controller, but it feels counter-intuitive for a `DirectoryView` instance to pass itself as an argument to `TreeViewPackage.updateIcon` and have `TreeViewPackage` inspect its properties before adding an icon

Reply:
Yeah, give it a try and see if we can simplify this interface to just be about assigning services and updating the icon with the best service. I think we're pretty close here.